expr/parse.go

```diff
@@ -115,12 +115,27 @@ type MetricRequest struct {
 	Until  int32
 }
 
+// ParseContext includes contextual information that affects how the current
+// expression is parsed.
+// This should only be set by internal calls.
+type ParseContext struct {
+	// Piped means the expression to be parsed is known to receive a piped
+	// input.
+	Piped bool
+	// IsFullArg means the expression to be parsed is a full argument to a
+	// function.
+	// This is used to work out whether we should attempt to parse the
+	// expression as a numerical constant - we should only do so if it's an
+	// full argument.
+	IsFullArg bool
```

Comment:
i'm having some trouble understanding what this means.
Do we ever parse "half" or "non full" args? when? what does it mean to be "full" vs "not full" ?
What does "Arg" mean?  for the expression `1 | fn()`, is 1 an arg? - this is equivalent to `fn(1)` so I think yes?

Reply:
An example might make the intention of it clearer: 

Given a query like: `funcA(1 | funcB(2))`:
* `1` should be parsed as a metric name. `1` is contained within the argument to `funcA`, but it's not the "full" argument. The full argument is `1 | funcB(2)`. 
* The `2` within `funcB` should be parsed as a number. This is because it doesn't have any pipes following it.

I see the name is not easy to understand though, maybe`CanParseAsNumber` would be better.

Reply:
I think what you call "full argument" is what most people would refer to as "the string containing the argument list" or "all arguments".
E.g. we have an existing function called parseArgList whose doc says " parseArgList parses a comma separated list of arguments"

Reply:
I'm trying to refer to a single argument in the function call, though struggling to come up with a good name.

If you have the function `funcA(1,2,3)`, `1`, `2` and `3` are separate "full" arguments (that should be parsed as numbers). 

For `funcA(1 | funcB(2), 3)`, `1` is a "partial" argument since there are additional characters until the comma indicating the end of one argument is reached.

Reply:
ah i misunderstood you earlier.  i think the terminology is OK then, just come comments will do.

Reply:
I renamed the flag to `CanParseAsNumber`, and added more details about when something is being parsed as a number to the comment in https://github.com/grafana/metrictank/pull/2012/commits/21418b79494179b1834262ad51ef2595787b17e2